perf: Reduce unnecessary FSDataOutputStream#hsync to enhance append performance

TheR1sing3un opened the pull request:
Before this optimization：
The average throughput on one of our real writer is 127k：
<img width="2424" height="596" alt="image" src="https://github.com/user-attachments/assets/4b6a872b-12e7-4165-8519-aaf94aac6e3a" />
After this optimization:
160k!
<img width="2432" height="602" alt="image" src="https://github.com/user-attachments/assets/4337a459-0983-4b07-9bb7-280c5caf0937" />
We have conducted stress tests in multiple scenarios, and the performance can be significantly increased by 20% to 40%


### Describe the issue this Pull Request addresses

closes #17516 

### Summary and Changelog

1. Reduce unnecessary `FSDataOutputStream#hsync` to enhance append performance

### Impact

write performance improvement

### Risk Level

low

### Documentation Update

none

### Contributor's checklist

- [x] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [x] Enough context is provided in the sections above
- [x] Adequate tests were added if applicable

TheR1sing3un committed:
perf: Reduce unnecessary `FSDataOutputStream#hsync` to enhance append performance

1. Reduce unnecessary `FSDataOutputStream#hsync` to enhance append performance

Signed-off-by: TheR1sing3un <chaoyang@apache.org>

TheR1sing3un committed:
feat: flush behavior compatible with the block append mode

1. flush behavior compatible with the block append mode

Signed-off-by: TheR1sing3un <chaoyang@apache.org>

yihua commented on hudi-hadoop-common/src/main/java/org/apache/hudi/common/table/log/HoodieLogFormatWriter.java:
```diff
@@ -244,21 +250,28 @@ public void close() throws IOException {
+    if (sync) {
+      // NOTE : the following API call makes sure that the data is flushed to disk on DataNodes (akin to POSIX fsync())
+      // See more details here : https://issues.apache.org/jira/browse/HDFS-744
+      output.hsync();
```
Is there a reason to do `hsync` for correctness @balaji-varadarajan-ai ?

yihua commented on hudi-hadoop-common/src/main/java/org/apache/hudi/common/table/log/HoodieLogFormatWriter.java:
```diff
@@ -245,6 +246,10 @@ public void close() throws IOException {
+      // Before closing the stream, call `FSDataOutputStream#hsync` to manually request DataNodes to perform data flushing to enhance the data persistence capability
+      // NOTE : the following API call makes sure that the data is flushed to disk on DataNodes (akin to POSIX fsync())
+      // See more details here : https://issues.apache.org/jira/browse/HDFS-744
+      output.hsync();
```
That's right — the key behavioral change is that `hsync` moves from per-`appendBlocks()` to per-file (on `closeStream()` and rollover). The `closeStream()` path now explicitly calls `flush(true)` to force `hsync`, so data is guaranteed to be persisted before a file is closed or rolled over. The window where data sits only in OS buffers is between successive `appendBlocks()` calls within the same log file, which is the intended tradeoff for the 20-40% throughput gain.

TheR1sing3un commented:
@hudi-agent review it

danny0405 commented on hudi-hadoop-common/src/test/java/org/apache/hudi/common/functional/TestHoodieLogFormat.java:
```diff
@@ -215,7 +215,7 @@ public void testEmptyLog() throws IOException {
   public void testBasicAppend(HoodieLogBlockType dataBlockType) throws IOException, InterruptedException, URISyntaxException {
     Writer writer =
-        HoodieLogFormat.newWriterBuilder().onParentPath(partitionPath)
+        HoodieLogFormat.newWriterBuilder().onParentPath(partitionPath).withSyncDuringFlush(true)
```
if it is okay we also call the hsync once during `closeStream` for 0.x table, can we add an explicit hsync api on the writer just for the tests? I see the tests are logging `auto-flushed` keywords, we might remove the auto-flush totally since it impacts the perf. We just need to ensure visibility for commit level to downstream instead of per log block.

TheR1sing3un replied:
> if it is okay we also call the hsync once during `closeStream` for 0.x table, can we add an explicit hsync api on the writer just for the tests? I see the tests are logging `auto-flushed` keywords, we might remove the auto-flush totally since it impacts the perf. We just need to ensure visibility for commit level to downstream instead of per log block.

Nice suggestion. done~

TheR1sing3un committed:
fixup: address review - drop syncDuringFlush, expose explicit sync()

Following @danny0405's suggestion in the PR review, ensure only
commit-level visibility on the production path:

- Remove the `withSyncDuringFlush` builder option and the
  `flush(boolean)` overload on HoodieLogFormatWriter; the production
  path no longer flushes or hsyncs at appendBlocks.
- Expose `Writer#sync()` (flush + hsync) as an explicit API for tests
  that assert per-append visibility on the underlying file system.
- closeStream still calls sync() once before close so a closed writer
  guarantees data is persisted to DataNodes.
- Update tests that previously relied on `withSyncDuringFlush(true)`
  to call `writer.sync()` explicitly before per-append FileStatus
  size assertions, and rename the related assertion message to drop
  the misleading "auto-flushed" wording.
